### PR TITLE
feat(resolver): support generic constraint using CEL

### PR DESCRIPTION
**Pull request by dinhxuanvu**
Signed-off-by: Vu Dinh <vudinh@outlook.com>

**Description of the change:**

Introduce the new type of constraint that uses CEL (Common Expression Language)
as an expression language to define the constraint in a generic way.
This type of constraint will allow operator authors to specify a dependency
on any arbitrary properties in the bundle. This constraint also supports
logical operator such as AND and OR in the CEL expression.

**Motivation for the change:**

https://github.com/operator-framework/enhancements/blob/master/enhancements/generic-constraints.md

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (perdasilva) [inline pkg/controller/registry/resolver/cache/predicates.go:345]: should we at least log that there was an unmarshalling error? Or, maybe panic?

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (perdasilva) [inline pkg/controller/registry/resolver/cache/predicates.go:355]: same here: should we log or panic?

File: pkg/controller/registry/resolver/constraints/cel.go
Comment (perdasilva) [inline pkg/controller/registry/resolver/constraints/cel.go:62]: it would be good to get some contextual comments to undrestand which options you are a setting and why (maybe with links to external CEL documentation?) (same for the `CompileOptions` method above)

File: pkg/controller/registry/resolver/constraints/cel.go
Comment (perdasilva) [inline pkg/controller/registry/resolver/constraints/cel.go:46]: should we call it `celProgram` or `program` instead of `p`? Single letter names make the code harder to read.

**Comment by dinhxuanvu** [comment]
Due to the automatic approved label added by the bot, I place a hold on here to avoid accident merging.
/hold

File: pkg/controller/registry/resolver/constraints/cel.go
Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/constraints/cel.go:62]: Will add some comments to explain some context here.

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/cache/predicates.go:355]: In the original code, it was a panic. However, as you can see other predicates don't panic on json unmarshal error as we simply ignore it. So that's why I do it this way. We currently don't have logger for predicates so there is no logging for that either.

Comment (perdasilva) [reply pkg/controller/registry/resolver/cache/predicates.go:355]: How hard would it be to add logging? If we don't want to handle this in this PR, can we track this somewhere? No logging could lead to some hard to find bugs. Especially with a generic constraints evaluator, being in the dark here could be a bit painful.

Comment (perdasilva) [reply pkg/controller/registry/resolver/cache/predicates.go:355]: I'd even suggest adding the panic back until we have logging

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (joelanford) [inline pkg/controller/registry/resolver/cache/predicates.go:337]: Maybe I'm being naive, but I'm a little surprised to see constraint-specific logic on the OLM side of this. I was imagining that we'd define a generic `Constraint` interface in `o-f/api` that `Cel`, `All`, `Any`, `None`, etc. would all implement.

That way, all OLM would need to do would be unmarshal the constraints and then use that generic interface to evaluate them.  What am I missing?

File: pkg/controller/registry/resolver/constraints/cel.go
Comment (joelanford) [inline pkg/controller/registry/resolver/constraints/cel.go:1]: Why is this code duplicated in both API and OLM? Shouldn't OLM just import from API?

File: pkg/controller/registry/resolver/resolver.go
Comment (joelanford) [inline pkg/controller/registry/resolver/resolver.go:357]: Use the `Constraint` struct from API here?

File: pkg/controller/registry/resolver/constraints/cel.go
Comment (perdasilva) [inline pkg/controller/registry/resolver/constraints/cel.go:118]: Should we also add a test for this code path?

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/cache/predicates.go:337]: This is predicate information. The `rule` and `message` are there to surface those information to the resolver during failure with `.String()` function. It is not the same as `Constraint` struct in `api` repo.

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/cache/predicates.go:345]: This is `continue` behavior is consistent with other predicates with the intention to ignore bad properties. Usually, bad properties shouldn't even make it to the cluster as bundle validation should catch this. A panic would cause the entire resolver action to fail which essentially means one bad property in one bundle will cause OLM to stop working entirely. I can go with `panic` for now given there is no logging here.

File: pkg/controller/registry/resolver/constraints/cel.go
Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/constraints/cel.go:1]: The PR on api repo was created later after I was adding some validation on registry repo and realized there were some shared codes between OLM and registry that can be consolidated in api repo.
I'm keeping the code here for now so that all of tests can be run properly while waiting for the api PR to merge. Even if we merge this PR first, I can do a subsequent PR to vendor new codes from api. That's merely a refactoring PR.

File: pkg/controller/registry/resolver/resolver.go
Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/resolver.go:357]: Yes. That's the intention after the api PR merged.

File: pkg/controller/registry/resolver/constraints/cel.go
Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/constraints/cel.go:118]: I saw your comment on this in api PR. Will address this over there given this code is moving over to api repo anyway.

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (perdasilva) [reply pkg/controller/registry/resolver/cache/predicates.go:345]: Yeah, that makes sense. Silent errors aren't the best tho. Is it very difficult to have attach a logger to the class? Maybe just instantiate a global one for the package? I'm not sure how logging works in Go and our projects yet. If it's too much effort or too outside the scope of the PR and panic can be disruptive to the user, I'd suggest creating a tracking issue and link it in a comment.

Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/cache/predicates.go:345]: I think logging the error is the better solution here but as you said, it is not in-scope of this PR to rework the predicate to include logging. The `panic` solution is more disruptive unfortunately. It is kind of lose-lose situation. At best, we can keep it status quo with the `continue` hence my original code. Anyway, let me see what I can do here.

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (anik120) [inline pkg/controller/registry/resolver/cache/predicates.go:360]: Same comment here, `CreateGenericPredicate`

**Review by anik120 — CHANGES_REQUESTED** [review]

File: pkg/controller/registry/resolver/cache/predicates.go
Comment (dinhxuanvu) [reply pkg/controller/registry/resolver/cache/predicates.go:334]: As we did discuss a bit on api [PR](https://github.com/operator-framework/api/pull/202#issuecomment-989401668), we defer any further interface design to a later time given right now given there isn't a need yet for a generic interface that can support multiple types of constraints. The generic constraint only supports CEL so it is a CEL constraint. We may or may not expand the support to other languages anytime soon.

**Comment by estroz** [comment]
It would be nice to refactor some of this into a more generic setup like in #2418 but I can always rebase my PR if it goes in after this one, or do a follow-up if this goes in last.

/lgtm

**Review by estroz — COMMENTED** [review]
Nice work @dinhxuanvu 

/lgtm

**Pull request merged**
